### PR TITLE
Fix bad routing during logout

### PR DESCRIPTION
**Pull request by bthooper**
I went down a number of rabbit holes on this one so I made more changes than I needed to in the end.

There is still an issue with our routes - since we are using Devise there is no  User Controller.  Thus the nested routes for the food resources creates routes for users which will generate a routing error.

However, the larger issue was that the rails app was not generating a DELETE request but a GET request.  Rails is supposed to use JavaScript to convert GETs to DELETES when told to do so but for some reason this stopped working.  I added a link to the head to the applicaiton.html.erb to include the needed javascript and another line in config/initializers/assets.rb to compile the relevant JS. I'm not sure if  this is the preferred way to accomplish this task...but that's what i did and it worked.

### CODE REVIEW

**Comment by thomascchen** [comment]
@bthooper it looks like Hotwire and Devise don't play well yet. See this issue: https://github.com/heartcombo/devise/issues/5358

Looks like a few people outlined a workaround there, might be worth giving that a shot: https://github.com/heartcombo/devise/issues/5358#issuecomment-798796788

**Comment by bthooper** [comment]
OK! This will be today's project then....I'll see what I can learn and report back.

**Comment by bthooper** [comment]
hmmm....so as i look back in to this...

I got it working by making sure the javascript was loading. i had played around with the button solution as described above, and turning off turbolinks...but with the javascript loaded it seemed to work and not require those work arounds....

So, I'm wondering if this needs to be done differently (eg as described in the devise forum comment) or if my solution is good???  Does this conflict with what you were doing with hotwire? I guess part of my problem is I don't know how we are using hotwire and what it is doing so i can't tell if im causing a problem.

**Comment by thomascchen** [comment]
@bthooper yea, I think Hotwire has its own way of doing JavaScript, which involves Turbo. Would be neat to do things the Hotwire/Turbo way (can read more about this at https://hotwire.dev) but if it's too much of a pain, we could strip out Turbo. It's been a while since I looked at this PR, so I think I need to refresh my memory. Will try to take a closer look as soon as I can.

**Comment by bthooper** [comment]
no worries...in the mean time i'll explore hotwire and see if i can come with a strategy there....

**Comment by bthooper** [comment]
More research here:
https://www.youtube.com/watch?v=yZDTBItc3ZM

This seems to walkthrough a solution directly related to our situation...that said, it's a bit beyond my experience and knowledge....so I'm willing to give it a try...but it might be ugly.

**Comment by thomascchen** [comment]
GoRails is a great resource! And that video looks... fairly recent? Hotwire is new to me too :) Thought this would be a good opportunity for us to learn! Give it a shot, and let's learn together!

**Comment by bthooper** [comment]
pushed an update based on gorails and additional research. bascially had to configure hotwire and devise - so i may have messed up with your hotwire implentation..but i'm still wrapping my head around hotwire....see you tonight?!?

**Comment by bthooper** [comment]
> pushed an update based on gorails and additional research. bascially had to configure hotwire and devise - so i may have messed up with your hotwire implentation..but i'm still wrapping my head around hotwire....see you tonight?!?

Perhaps not tonight? Looks like it might get cancelled....happy Memorial Day!

File: Gemfile [hunk 1]
```diff
@@ -25,6 +25,8 @@ gem 'jbuilder', '~> 2.7'
 #
 # Use Devise for authentication
 gem 'devise'
+# Add jquery
+gem 'jquery-rails'
```
Comment (thomascchen) [inline Gemfile]: Hmm... let's avoid adding jQuery if we can. jQuery is not really needed as much given modern JavaScript's advancements in recent years. And I've love for us to lean into Stimulus for JS needs in this app as that is the direction Rails is headed...

File: Gemfile [hunk 2]
```diff
@@ -59,3 +61,5 @@ end
 
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
 gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
+
+gem "hotwire-rails", "~> 0.1.3"
```
Comment (thomascchen) [inline Gemfile]: Let's add `gem "stimulus-rails"` under line 15 instead of this gem. "hotwire-rails" basically bundles together "stimulus-rails" and "turbo-rails". Since they are basically standalone gems, I'd love to install them separately.

File: app/javascript/controllers/hello_controller.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { Controller } from "stimulus"
```
Comment (thomascchen) [inline app/javascript/controllers/hello_controller.js]: I think we can delete this controller?

File: app/javascript/packs/application.js [hunk 4]
```diff
@@ -3,10 +3,11 @@
 // a relevant structure within app/javascript and only use these pack files to reference
 // that code so it'll be compiled.
 
-import { Turbo, cable } from '@hotwired/turbo-rails';
-import * as ActiveStorage from '@rails/activestorage';
-import 'channels';
-import 'controllers';
+import { Turbo, cable } from "@hotwired/turbo-rails";
```
Comment (thomascchen) [inline app/javascript/packs/application.js]: We can probably delete this import on line 6, since it's essentially duplicated on line 7. Line 7 seems to be the preferred import approach now.

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -6,10 +6,9 @@
 
   <meta charset="UTF-8">
   <meta name="viewport" content="width=device-width, initial-scale=1">
-
   <%= csrf_meta_tags %>
   <%= csp_meta_tag %>
-
+  <%= javascript_include_tag "rails-ujs" %>
```
Comment (thomascchen) [inline app/views/layouts/application.html.erb]: Hotwire Turbo is basically the new approach that is replacing UJS (unobtrustive javascript), so we shouldn't re-introduce this here...

File: config/initializers/assets.rb [hunk 6]
```diff
@@ -11,4 +11,5 @@
 # Precompile additional assets.
 # application.js, application.css, and all non-JS/CSS in the app/assets
 # folder are already added.
-# Rails.application.config.assets.precompile += %w( admin.js admin.css )
+
+Rails.application.config.assets.precompile += %w( rails-ujs.js )
```
Comment (thomascchen) [inline config/initializers/assets.rb]: let's undo this change. I don't think we want to reintroduce UJS...

File: config/initializers/devise.rb [hunk 7]
```diff
@@ -1,4 +1,33 @@
 # frozen_string_literal: true
+#
+class TurboFailureApp < Devise::FailureApp
```
Comment (thomascchen) [inline config/initializers/devise.rb]: So I watched the GoRails video, and I think it's out of date already at this point unfortunately. I think we can actually skip over all of this custom controller approach and configuration changes in favor of a much simpler solution...

File: config/routes.rb [hunk 8]
```diff
@@ -13,4 +13,9 @@
   resources :users do
     resources :food_access_points, only: [:create, :new, :edit, :update, :destroy]
   end
+  
+  # resources :user do
+  #   resources :food_access_points, only: [:create, :new, :edit, :update, :destroy]
+  # end
```
Comment (thomascchen) [inline config/routes.rb]: I think we can remove this commented out stuff right?

File: config/routes.rb [hunk 9]
```diff
@@ -13,4 +13,9 @@
   resources :users do
```
Comment (thomascchen) [inline config/routes.rb]: when I was poking around here, I realized we haven't created a `UsersController` yet, which Rails was complaining about. I think for now we need to tell Rails to skip over any users routes that it might be expecting by specifying `only: []`. That will tell Rails not to except any of the routes for showing/editing/deleting/etc. users until we're ready to add them.

```suggestion
  resources :users, only: [] do
```

**Review by thomascchen — CHANGES_REQUESTED** [review]
This is a great start! I've left a few comments here, but I think there are two main causes to the buggy behavior we're seeing:

1) We never created a UsersController, which Rails is expecting, when we define the users nested routes here:

```ruby
  resources :users do
    resources :food_access_points, only: [:create, :new, :edit, :update, :destroy]
  end
```

If you run `rails routes` you'll see the following:
Screen Shot 2021-06-05 at 4 23 39 PM

and when we sign out, Devise seems to be looking for one of those routes, which we haven't defined. We actually don't need those right now, so we can add `only: []` to those users routes to skip all of those routes. That handles that `UsersController` error that we've been seeing when trying to sign out.

2) The second issue is with Hotwire Turbo. Turbo (and formerly UJS) basically takes over the normal request/response behavior through JavaScript to make page transitions happen more seamlessly. This works fine for GET requests, but doesn't work great for POST/PATCH/DELETE requests. When we need to handle those types of actions now with existing `link_to`s, we need to turn Turbo off for those, by passing in `data: { turbo: false }`, which lets Devise behave as it used to. We should also turn those `link_to`s into `button_to`s per this guidance: https://github.com/hotwired/turbo-rails/#compatibility-with-rails-ujs

So we need to make this change in two places:
https://github.com/codeforUV/food_resources/blob/master/app/views/application/_header.html.erb#L44

and 

https://github.com/codeforUV/food_resources/blob/master/app/views/application/_header.html.erb#L68

those should both become:

```ruby
<%= button_to "Sign Out", destroy_user_session_path, method: :delete, data: { turbo: false }, class: "<all the classes...>" %>
```

Once we make those changes, I think that should actually fix our situation, and we can undo a bunch of the specific config changes you made here...

I have this working locally, let me know if you'd like me to push it up or we can find a time to pair on this sometime!

File: Gemfile (hunk 1 above)
Comment (bthooper) [reply Gemfile]: roger...
i was exploring the problem with the user controller and this was part of an earlier effort to fix it...ujs is gone!

File: config/routes.rb (hunk 9 above)
Comment (bthooper) [reply config/routes.rb]: OK...so I adopted your suggestions as best I could.

There is still a problem.  The delete button works as expected (but now it's an ugly button so I'll need to fix the styling).  However I can not get error messages to show on the devise related forms.  That still seems like a problem with turbo and devise...and I'm not sure how to address it.

I would love a coding night! Or just shoot me in the right direction. I started a simple demo app using hotwire, to try to wrap my head around it a bit...still playing with that.

Comment (thomascchen) [reply config/routes.rb]: Awesome, I think this looks good! I merged in some dependency updates to the master branch, so it looks like you need to either merge master into this branch or rebase with master to get up to date with those changes, then just run bundle install && yarn, and we should be good to merge!

Once those merge conflicts have been addressed, I’d be in favor of merging this as is and then doing the styling and digging into those error messages in new PRs.

I’ll take a look at my calendar and try to figure out another coding night, will hit you up on Slack!

Comment (bthooper) [reply config/routes.rb]: i gotta get a job so i use git all the time and better understand what I am doing!!

In any event, I think I have merged master and pushed up the changes....i think....i hope...

File: Gemfile.lock [hunk 10]
```diff
@@ -254,8 +257,9 @@ DEPENDENCIES
   jbuilder (~> 2.7)
   listen (~> 3.2)
   pg (~> 1.0)
-  puma (~> 4.3)
-  rails (~> 6.1.2)
+  puma (~> 4.1)
```
Comment (thomascchen) [inline Gemfile.lock]: i think we want this to remain 4.3

Comment (thomascchen) [reply Gemfile.lock]: when i bundle install locally i also get a `stimulus-rails` added on line 270. can you try running bundle install on this again and commiting those changes?

**Review by thomascchen — CHANGES_REQUESTED** [review]
a couple minor dependency issues! almost there!

File: Gemfile.lock (hunk 10 above)
Comment (bthooper) [reply Gemfile.lock]: ok....tried again.....

**Review by thomascchen — APPROVED** [review]
🎉
